plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/DetachableBeanHolder.java

```diff
@@ -65,30 +62,55 @@ public DetachableBeanHolder(Object bean, Object beanFactry, Class<?>[] paramClas
         this.beanFactory = beanFactry;
         this.paramClasses = paramClasses;
         this.paramValues = paramValues;
-        beanProxies.add(new WeakReference<DetachableBeanHolder>(this));
+
+        String beanName = deduceBeanName(beanFactry, paramClasses, paramValues);
+        if (beanName != null) {
+            synchronized (beanProxies) {
+                beanProxies.computeIfAbsent(beanName, k -> new ArrayList<>());
+                beanProxies.get(beanName).add(new WeakReference<>(this));
+            }
+        }
+    }
+
+    protected static String deduceBeanName(Object beanFactory, Class<?>[] paramClasses, Object[] paramValues) {
```

Comment:
If the first argument is always bean name, then the logic of 'deduceBeanName' can be simplified.